### ISSUE TITLE
Ending task in Task Scheduler in Windows does not stop Syncthing from running

### ISSUE THREAD

**Issue by tomasz1986** [syncthing/syncthing#7042]
When using the recommended way to set Syncthing up to automatically start and run in background in Windows with Task Scheduler following the official guide under https://docs.syncthing.net/users/autostart.html#autostart-windows-taskschd, ending the task does not stop Syncthing from running in the background.

The reason is that Syncthing spawns a sub-process, so that there are two `syncthing.exe` processes running at the time. The higher one is the monitor process (see https://docs.syncthing.net/users/syncthing.html#cmdoption-no-restart), and only this one is being controlled by Task Scheduler. Thus, ending the task only kills the monitor process, but leaves the main one untouched.

I do not think that there is any solution to the problem right now, as running Syncthing with the `-no-restart` option would prevent automatic updates and other panic recoveries, which I do not think is desirable. On the other hand, the current state makes it impossible to simply stop Syncthing from running using the Task Scheduler, which is also not good from the user's perspective, i.e. the user thinks that they have made Syncthing exit, while in reality the synchronisation is still going on invisibly in the background.

I will think of a possible solution using the Task Scheduler, but with the current monitor-main process construct, there may be none.

**Comment by imsodin** [comment]
`-no-restart` does not control if there is a monitor process since quite a while: There always is a monitor process (e.g. for panic handling). And the monitor stops the child process when it gets terminated, so it should work. Looks like whatever the Task Scheduler does to terminate the process does not result in a `os.Interrupt` signal, but directly terminates the process (equivalent of SIGKILL).  
Maybe the culprit is the option "If the running task does not end when requested, force it to stop" and the scheduler is just too impatient when stopping. Could you test if unchecking that helps please.

**Comment by tomasz1986** [comment]
> Maybe the culprit is the option "If the running task does not end when requested, force it to stop" and the scheduler is just too impatient when stopping. Could you test if unchecking that helps please.

There is no difference, unfortunately. I have just tried, and only the monitor process gets killed regardless of the setting.

**Comment by imsodin** [comment]
> I guess then that the information from https://docs.syncthing.net/includes/env-vars.html?highlight=stnorestart is outdated also, right?

Yes it is.

I don't think there's anything we can do here. The best info I found on this says it's not possible and suggest using `taskkill` instead: https://superuser.com/questions/959364/on-windows-how-can-i-gracefully-ask-a-running-program-to-terminate  
A quick search also didn't turn up any option to run a command on stop (to trigger shutdown through the API). Looks like it's simply not capable to gracefully terminate a task. We should probably add a warning about that to the docs.

**Comment by calmh** [comment]
The code required to support running as a Windows service isn't enormous; it's basically a couple of start/stop callbacks and there are Go packages for it. Might be an option to implement that ourselves and recommend running as a service instead.

**Comment by tomasz1986** [comment]
> The best info I found on this says it's not possible and suggest using `taskkill` instead

Yeah, it would be nice to be able to run custom commands on task end, but there is no such functionality in Task Scheduler.

> We should probably add a warning about that to the docs.

I will try to add the warning sometime this week.

> Might be an option to implement that ourselves and recommend running as a service instead.

Service is problematic though. It would need to be run as a per-user service (see https://docs.microsoft.com/en-us/windows/application-management/per-user-services-in-windows), which is available only in Windows 10 and up. Otherwise, a normal service runs system-wide, which is not suitable for Syncthing, unless we are talking about a single-user machine. It is also impossible to control services without administrative access, unless you fiddle around with service permissions, doing which also requires administrative access to begin with :wink:.

A built-in service functionality would be a replacement for the current NSSM solution though.

**Comment by calmh** [comment]
I think "unless it's a single user machine" covers a very large majority of Windows boxes running Syncthing, and the rest are managed by people who understand the nuances. But it's a side track anyway.

**Comment by imsodin** [comment]
I took the liberty of repurposing this issue as there is no bug and opened an issue on the docs repo: https://github.com/syncthing/docs/issues/577

**Comment by Catfriend1** [comment]
I've also had trouble with the monitor process. I'm using Syncthing as a service wrapped by nssm under LocalService account since a long time. However, nssm has the weakness not being configurable to react on Syncthing's exit codes. To solve problematic "monitor respawns child when child was killed" and "upgrade exits instead of respawning a child process because monitor exits and nssm sees it" I've made a batch script to mimick the Syncthing monitor process behaviour suitably for nssm. ( https://github.com/Catfriend1/syncthing-msi/blob/v1.8.0/Syncthing/syncthing_service.cmd ) nssm starts this script instead of Syncthing and upgrade, restart or user shutdown behaviour is all fine now. Maybe this could also help task scheduler users?

**Comment by ProactiveServices** [comment]
> However, nssm has the weakness not being configurable to react on Syncthing's exit codes. To solve problematic "monitor respawns child when child was killed" and "upgrade exits instead of respawning a child process because monitor exits and nssm sees it"

It's been a while since I've run Syncthing on Windows, but the method outlined on step 12 [here](https://docs.syncthing.net/users/autostart.html#autostart-windows-tools) was always reliable for me during upgrades and graceful shutdowns via Syncthing itself.

**Comment by Catfriend1** [comment]
@ProactiveServices cool info, thanks. I didn't know about the nssm cli to configure exit code specific behaviour 👍

**Comment by Magissia** [comment]
Honestly, nssm/FireDaemon both do the job.

**Comment by shoogle** [comment]
> Syncthing spawns a sub-process, so that there are two syncthing.exe processes running at the time. [...] Ending the task only kills the monitor process, but leaves the main one untouched.

Could the main process periodically check to see if the monitor process is still running, and if not, exit gracefully?

**Comment by Bill-Stewart** [comment]
I wrote a Windows installer package that installs a Windows service if you install for all users:

https://github.com/Bill-Stewart/SyncthingWindowsSetup

It uses NSSM and runs using the `LOCAL SERVICE` account, so you have to grant "Modify" permissions for this account to directories you want to add to your configuration.

**Comment by michaelsmoody** [comment]
@Bill-Stewart I just want to give me approval, thumbs-up, and thanks for this. While I'm not actively using it at the moment, I could _easily_ see this being added to Chocolatey, or another similar package manager, and become even the de-facto installer for Syncthing from "other than official" sources. Quite frankly, it's probably long past time that this works this way.

Thank you again for your efforts, and this is good work (I say this after a reasonably quick but "in depth quick" review of your repo).

**Comment by Bill-Stewart** [comment]
Thanks for the review @michaelsmoody . Just an update that the installer I built now uses a local user account rather than `LOCAL SERVICE` (the security reason is that `LOCAL SERVICE` is a member of the local `Users` group, and this is too much access).

**Comment by michaelsmoody** [comment]
> `and this is too much access)`

I understand, and generally completely agree. I might suggest a configuration item to allow the user to choose. There are situations where I might need **LOCAL SERVICE** for FACL reasons (I recently had a situation exactly like this). As well as possibly items such as IIS_IUSR, etc. As rare as these edge cases are, I could see it occurring, as I have had need to do exactly this on Linux-based systems (the user/group I run Syncthing as, dealing with network drives, and have had to working with some......_odd_ scenarios ¯\_(ツ)_/¯)

However, once again, I compliment you on your excellent choice of details.
